### PR TITLE
Python Package updates

### PR DESCRIPTION
**Pull request by JustinElms**
## Background
In an effort to boost Navigator performance and UX we decided to upgrade several python packages during our Dec 6-7th code sprint. We initially updated the packages listed in the table below and found a significant performance gain - the time to complete our profiling test suite dropped from 741 seconds to 335! A new environment based on these packages can be created using the command `conda env create -f=config/conda/environment.yml`. 

So far, only minor changes were needed to get the Navigator running in this environment which have been addressed in the initial commit. This PR is meant to be used to find and address any other issues that arise. To test this branch I'm using the UI testing spread sheet used for release checks which can be found at: 
https://docs.google.com/spreadsheets/d/1oRlgB9MTyVtvCyj7-r2aUz9zOFCymy_Kn8ccMcxmjy0/edit?usp=sharing

Package | Previous Version | Current Version
-- | -- | --
python | 3.6.8 | 3.8.12
basemap | 1.2.0 | 1.2.2
basemap-data-hires | 1.2.0 | 1.2.2
blinker | 1.4 | 1.4
cachetools | 2.0.1 | 4.2.4
cloudpickle | 1.2.0 | 2.0.0
cmocean | 2 | 2
dask | 1.2.2 | 2021.11.2
flask | 0.12.2 | 2.0.2
flask-babel | 0.11.2 | 2.0.0
flask-compress | 1.4.0 | 1.10.1
flask-sqlalchemy | 2.4.1 | 2.5.1
gdal | 2.3.3 | 3.3.2
geojson | 2.5.0 | 2.5.0
geopy | 1.13.0 | 2.2.0
gunicorn | 20.0.4 | 20.1.0
marshmallow | 3.11.1 | 3.14.1
matplotlib | 3.1.0 | 3.5.0
metpy | 0.10.2 | 1.1.0
netcdf4 | 1.5.1.2 | 1.5.8
numpy | 1.16.2 | 1.21.4
pandas | 0.24.2 | 1.3.4
pip | 19.1.1 | 21.3.1
ply | 3.11 | 3.11
pykdtree | 1.3.1 | 1.3.4
pyresample | 1.9.0 | 1.22.2
requests | 2.22.0 | 2.26.0
scikit-image | 0.15.0 | 0.18.3
scipy | 1.1.0 | 1.7.3
seawater | 3.3.4 | 3.3.4
sentry-sdk | 0.19.3 | 1.5.0
sqlalchemy | 1.3.13 | 1.4.27
werkzeug | 0.14.1 | 2.0.1
xarray | 0.11.3 | 0.20.1
zarr | 2.6.1 | 2.10.3

## To Do
- [x] Create conda environment .spec file
- [x] UI Testing

## Checks
- [ ] I ran unit tests.
- [x] I've tested the relevant changes from a user POV.

_Hint_ To run all python unit tests run the following command from the root repository directory:
```sh
bash run_python_tests.sh
```

### CODE REVIEW

**Comment by JustinElms** [comment]
New deprecation warning for the updated pyproj package (upgraded from 1.9.5.1 to 3.3.0):
`
/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/pyproj/crs/crs.py:131: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
`
It looks like we need to get rid of the `init=` in our calls to Proj, ie. replace `Proj(init='EPSG:4326')` with `Proj('EPSG:4326')`. However, making this change causes the tiles to be placed randomly on the global map, probably because of the axis order change mentioned in the deprecation warning. 

Also, the [docs](https://pyproj4.github.io/pyproj/stable/api/index.html) state that  `pyproj.transform()` has been deprecated since version 2.6.1 and should be replaced with `pyproj.transformer()`. Strange that this still works and doesn't give us another warning.

**Comment by dwayne-hart** [comment]
Packages mysql-connector-c and  mysqlclient are required to support the Observation component of ONAV. However, when trying to get the data from a track the web services encounters an Internal Server Error.

| Package | Previous version | New version |
|:--- | :-----: | :---: |
| mysql-common | N/A | 8.0.27 | 
| mysql-connector-c | 6.1.11 | N/A |
| mysql-connector-python |   N/A | 8.0.27 |
| mysql-libs | N/A | 8.0.27 | 
|mysqlclient  | 1.4.6 |  2.0.3 |

```
[2021-12-10 14:01:03 -0330] [12476] [ERROR] Error handling request /api/v1.0/plot/?query=%7B%22dataset%22%3A%22giops_day%22%2C%22depth%22%3A0%2C%22endtime%22%3A%222021-12-11T00%3A00%3A00.000Z%22%2C%22latlon%22%3Afalse%2C%22quantum%22%3A%22day%22%2C%22showmap%22%3Atrue%2C%22starttime%22%3A%222020-01-01T00%3A00%3A00.000Z%22%2C%22track%22%3A%5B263%5D%2C%22track_quantum%22%3A%22month%22%2C%22trackvariable%22%3A%5B0%5D%2C%22type%22%3A%22track%22%2C%22variable%22%3A%5B%22votemper%22%5D%7D&format=json
Traceback (most recent call last):
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/gunicorn/workers/gthread.py", line 271, in handle
    keepalive = self.handle_request(req, conn)
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/gunicorn/workers/gthread.py", line 323, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/sentry_sdk/integrations/flask.py", line 90, in sentry_patched_wsgi_app
    return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/sentry_sdk/integrations/wsgi.py", line 140, in __call__
    reraise(*_capture_exception(hub))
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/sentry_sdk/_compat.py", line 54, in reraise
    raise value
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/sentry_sdk/integrations/wsgi.py", line 133, in __call__
    rv = self.app(
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/sentry_sdk/integrations/flask.py", line 90, in <lambda>
    return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/on/tools/miniconda/3/amd64/envs/navigator-2021.12/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/home/on/Ocean-Data-Map-Project/routes/api_v1_0.py", line 600, in plot_v1_0
    img, mime, filename = plotter.run()
  File "/home/on/Ocean-Data-Map-Project/plotting/plotter.py", line 57, in run
    _ = self.prepare_plot()
  File "/home/on/Ocean-Data-Map-Project/plotting/plotter.py", line 50, in prepare_plot
    self.load_data()
  File "/home/on/Ocean-Data-Map-Project/plotting/track.py", line 72, in load_data
    variables = [datatypes[int(x)] for x in self.trackvariables]
  File "/home/on/Ocean-Data-Map-Project/plotting/track.py", line 72, in <listcomp>
    variables = [datatypes[int(x)] for x in self.trackvariables]
IndexError: list index out of range
```

**Comment by JustinElms** [comment]
Not sure if this is an issue but I get this error when setting `processes=4` in runserver.py. Possible change to Flask or Gunicorn that needs to be addressed.

```
Traceback (most recent call last):
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/sentry_sdk/integrations/threading.py", line 69, in run
    reraise(*_capture_exception())
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/sentry_sdk/_compat.py", line 54, in reraise
    raise value
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/sentry_sdk/integrations/threading.py", line 67, in run
    return old_run_func(self, *a, **kw)
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/werkzeug/serving.py", line 948, in inner
    srv = make_server(
  File "/home/ubuntu/tools/miniconda/3/amd64/envs/nav-pkg-update/lib/python3.8/site-packages/werkzeug/serving.py", line 778, in make_server
    raise ValueError("cannot have a multithreaded and multi process server.")
ValueError: cannot have a multithreaded and multi process server.
```

File: plotting/map.py [hunk 1]
```diff
@@ -817,7 +817,8 @@ def plot(self):
 
         # Map Info
         self.basemap.drawmapboundary(fill_color=(0.3, 0.3, 0.3), zorder=-1)
-        self.basemap.drawcoastlines(linewidth=0.5)
+        if self.basemap.coastsegs[0]:  # ensure map contains coastline before trying to draw
```
Comment (JustinElms) [inline plotting/map.py:820]: Looks like Basemap 1.2.2 can't handle a call to `drawcoastlines` if the map doesn't intersect the ocean anywhere so we need some sort of check like this.

**Comment by JustinElms** [comment]
#



> Packages mysql-connector-c and mysqlclient are required to support the Observation component of ONAV. However, when trying to get the data from a track the web services encounters an Internal Server Error.
> 
> ....

@dwayne-hart I think this issue has been fixed with commits 239f311, e4904ae, and af66068. I believe mysqldb is depreciated now so I added mysqlclient to the environment to get the obs working again. This required me to install Python 3 and MySQL development headers using `$ sudo apt-get install python3-dev default-libmysqlclient-dev build-essential` which will have to documented for future Navigator installations. We can also look into using another library that works without this additional step such as  `mysql-connector-python`.

**Review by dwayne-hart — APPROVED** [review]

**Comment by JustinElms** [comment]
UI Testing complete, all functions work as expected!

**Pull request merged**